test/correctness/intrinsics.cpp

```diff
@@ -0,0 +1,199 @@
+    Expr f32y = Variable::make(Float(32, 4), "f32y");
+
+    // Check powers of two multiply/divide rewritten to shifts.
+    check(i8x * 2, i8x << 1);
+    check(u8x * 4, u8x << 2);
+    check(i8x / 8, i8x >> 3);
+    check(u8x / 4, u8x >> 2);
+
+    check(i16(i8x) * 4096, widening_shift_left(i8x, u8(12)));
+    check(u16(u8x) * 128, widening_shift_left(u8x, u8(7)));
+    //check(u32(u8x) * 256, u32(widening_shift_left(u8x, u8(8))));
+
+    // Check widening arithmetic
+    check(i16(i8x) + i8y, widening_add(i8x, i8y));
+    check(u16(u8x) + u8y, widening_add(u8x, u8y));
+    check(i16(u8x) + u8y, i16(widening_add(u8x, u8y)));
+    check(f32(f16x) + f32(f16y), widening_add(f16x, f16y));
+
+    check(i16(i8x) - i8y, widening_sub(i8x, i8y));
+    check(i16(u8x) - u8y, widening_sub(u8x, u8y));
+    check(f32(f16x) - f32(f16y), widening_sub(f16x, f16y));
+
+    check(i16(i8x) * i8y, widening_mul(i8x, i8y));
+    check(u16(u8x) * u8y, widening_mul(u8x, u8y));
+    check(i32(i8x) * i8y, i32(widening_mul(i8x, i8y)));
+    check(u32(u8x) * u8y, u32(widening_mul(u8x, u8y)));
+    check(f32(f16x) * f32(f16y), widening_mul(f16x, f16y));
+
+    // Widening mul allows mixed signs
+    check(i16(i8x) * u8y, widening_mul(i8x, u8y), Int(16, 4));
+    check(i32(i8x) * u8y, i32(widening_mul(i8x, u8y)));
+
+    // Excessive widening should be moved outside the arithmetic.
+    check(widening_mul(i16(i8x), i16(i8y)), i32(widening_mul(i8x, i8y)));
+    check(widening_mul(u16(u8x), u16(u8y)), u32(widening_mul(u8x, u8y)));
+    check(widening_mul(i16(i8x), i16(u8y)), i32(widening_mul(i8x, u8y)));
+    check(widening_mul(i16(u8x), i16(i8y)), i32(widening_mul(u8x, i8y)));
+    check(widening_mul(f32(f16x), f32(f16y)), f64(widening_mul(f16x, f16y)));
+
+    check(widening_add(i16(i8x), i16(i8y)), i32(widening_add(i8x, i8y)));
+    check(widening_add(u16(u8x), u16(u8y)), u32(widening_add(u8x, u8y)));
+    check(widening_add(i16(u8x), i16(u8y)), i32(widening_add(u8x, u8y)));
+    check(widening_add(f32(f16x), f32(f16y)), f64(widening_add(f16x, f16y)));
+
+    check(widening_mul(i32(i8x), i32(i8y)), i64(widening_mul(i8x, i8y)));
+    check(widening_mul(u32(u8x), u32(u8y)), u64(widening_mul(u8x, u8y)));
+    check(widening_mul(i32(i8x), i32(u8y)), i64(widening_mul(i8x, u8y)));
+    check(widening_mul(i32(u8x), i32(i8y)), i64(widening_mul(u8x, i8y)));
+
+    check(widening_add(i32(i8x), i32(i8y)), i64(widening_add(i8x, i8y)));
+    check(widening_add(u32(u8x), u32(u8y)), u64(widening_add(u8x, u8y)));
+    check(widening_add(i32(u8x), i32(u8y)), i64(widening_add(u8x, u8y)));
+
+    // Tricky case.
+    check(i32(u8x) + 1, i32(widening_add(u8x, u8(1))));
+
+    // Check saturating arithmetic
+    check(i8_sat(i16(i8x) + i8y), saturating_add(i8x, i8y));
+    check(u8_sat(u16(u8x) + u8y), saturating_add(u8x, u8y));
+    check(u8(min(u16(u8x) + u16(u8y), 255)), saturating_add(u8x, u8y));
+    //check(u8(min(i16(u8x) + i16(u8y), 255)), saturating_add(u8x, u8y));
```

Comment:
Why is this commented out? Looks like a valid saturating add to me.

Reply:
I brought it back. Matching it requires matching a cast in the pattern, which was stumping me earlier.